### PR TITLE
Add pre-commit hook config

### PR DESCRIPTION
**Pull request by stefmolin**
Addresses https://github.com/mwaskom/seaborn/issues/2950

I'm adding an initial version of a pre-commit config file for contributors to run pre-commit hooks. After testing it out, let me know if you want any changes and whether/where we should put the setup instructions for contributors in the docs.

---

Setup (from the top-level of the repo):
```shell
$ pip install pre-commit
$ pre-commit install
```

The next time you `git commit` the hooks will run on the staged files. If there were any changes to the versions of checks in the config file since the last commit or this is your first time using it, it will pull down the specified versions for you:

Screen Shot 2022-08-19 at 8 48 47 PM

If you don't pass all the checks, the commit doesn't happen, but you can tack on `--no-verify` to your `git commit` command to bypass the checks.

You can test the hooks on files without committing:
```shell
$ pre-commit run --file seaborn/palettes.py
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
mypy.................................................(no files to check)Skipped
```

Please note that for `flake8` the config option for `exclude` does not seem to be honored like the others in `setup.cfg`, so I added the equivalent pattern as an exclusion in the YAML file. Here, you can see that `seaborn/cm.py` is ignored for the `flake8` check:

```shell
$ pre-commit run --file seaborn/cm.py
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing seaborn/cm.py

trim trailing whitespace.................................................Passed
flake8...............................................(no files to check)Skipped
mypy.................................................(no files to check)Skipped
```

Use `--all-files` to run on the full repo:
```shell
$ pre-commit run --all-files
```

<details>
<summary>output</summary>
This yields some findings:

```
check yaml...............................................................Passed
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing doc/_templates/autosummary/base.rst
Fixing doc/_templates/autosummary/object.rst
Fixing ci/getmsfonts.sh
Fixing doc/nextgen/Makefile
Fixing seaborn/cm.py
Fixing doc/index.rst
Fixing doc/nextgen/index.rst
Fixing doc/citing.rst
Fixing seaborn/external/docscrape.py
Fixing README.md
Fixing doc/whatsnew/v0.11.1.rst
Fixing ci/deps_pinned.txt
Fixing doc/whatsnew/v0.2.0.rst
Fixing licences/NUMPYDOC_LICENSE

trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing doc/whatsnew/v0.2.0.rst
Fixing doc/_static/copybutton.js
Fixing README.md
Fixing doc/citing.rst
Fixing doc/whatsnew/v0.11.0.rst
Fixing doc/README.md
Fixing doc/index.rst
Fixing doc/whatsnew/v0.7.0.rst
Fixing doc/nextgen/index.rst

flake8...................................................................Failed
- hook id: flake8
- exit code: 1

tests/test_utils.py:430:11: E275 missing whitespace after keyword

mypy.....................................................................Passed
```
</details>

Additional hooks can be found [here](https://pre-commit.com/hooks.html), and the docs for pre-commit are [here](https://pre-commit.com).

### CODE REVIEW

**Comment by mwaskom** [comment]
Thanks @stefmolin I'll probably wait to engage with this until after the 0.12.0 release but it looks cool.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    -   id: end-of-file-fixer
+    -   id: trailing-whitespace
+        exclude: \.svg$
+-   repo: https://github.com/pycqa/flake8
+    rev: 5.0.4
+    hooks:
+    -   id: flake8
+        exclude: seaborn/(cm\.py|external/)
+        types: [file, python]
+-   repo: https://github.com/pre-commit/mirrors-mypy
```
Comment (mwaskom) [inline .pre-commit-config.yaml:15]: Just curious but what's the rationale for pointing at this repo vs the main mypy one?

Comment (stefmolin) [reply .pre-commit-config.yaml:15]: The mirrors-mypy repo provides a [hook](https://github.com/pre-commit/mirrors-mypy/blob/main/.pre-commit-hooks.yaml) that can be used out of the box, but the mypy repo itself doesn't do so (you would have to supply all the logic to run it in this YAML file).

**Comment by mwaskom** [comment]
Thanks @stefmolin for setting this up. TBH I don't love how you end up with two sources of truth on what will be enforced vs. checked, but it's limited enough in scope to be manageable.

Can you please rebase (so you pick up https://github.com/mwaskom/seaborn/pull/3009) and then add `pre-commit `to the `dev` extras in `pyproject.toml`.

If you want to add  a sentence in the README about needing to do `pre-commit install` that would make sense too, probably right after the bit about running `make lint`. (e.g., "Alternately, you can use pre-commit to run lint checks on any files you are committing ...")

Also it looks like there's a handful of whitespace things that get caught when running this on all files. Do you want to run that and then commit the result as part of this PR too?

**Comment by stefmolin** [comment]
> Thanks @stefmolin for setting this up. TBH I don't love how you end up with two sources of truth on what will be enforced vs. checked, but it's limited enough in scope to be manageable.

Do you say this because there are additional checks here compared to `make lint`? Or because the Makefile has testing commands in there and this doesn't?

> Can you please rebase (so you pick up #3009) and then add `pre-commit `to the `dev` extras in `pyproject.toml`.
> 
> If you want to add a sentence in the README about needing to do `pre-commit install` that would make sense too, probably right after the bit about running `make lint`. (e.g., "Alternately, you can use pre-commit to run lint checks on any files you are committing ...")
> 
> Also it looks like there's a handful of whitespace things that get caught when running this on all files. Do you want to run that and then commit the result as part of this PR too?

All done.

**Comment by mwaskom** [comment]
> Do you say this because there are additional checks here compared to make lint? Or because the Makefile has testing commands in there and this doesn't?

Oh I just mean that within the lint/type checks, there's some duplication of how files get selected for / excluded from testing.

**Comment by stefmolin** [comment]
> Oh I just mean that within the lint/type checks, there's some duplication of how files get selected for / excluded from testing.

I think I understand what you are saying. The files to include/exclude can be specified for [all hooks once at the top](https://pre-commit.com/index.html#pre-commit-configyaml---top-level), but the hooks don't completely overlap in what they are looking at just like the `flake8` config in the [setup.cfg](https://github.com/mwaskom/seaborn/blob/master/setup.cfg#L3) file and the `mypy` config in the [Makefile](https://github.com/mwaskom/seaborn/blob/master/Makefile#L13) are different.

**Review by mwaskom — APPROVED** [review]

**Comment by mwaskom** [comment]
Thanks @stefmolin, may continue to tweak over time but expect this to be very helpful!

**Pull request merged**